File: app/src/main/java/org/descartae/android/view/fragments/facility/FacilitiesFragment.java
package org.descartae.android.view.fragments.facility;

import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.afollestad.materialdialogs.MaterialDialog;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapFragment;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;

import org.descartae.android.DescartaeApp;
import org.descartae.android.FacilitiesQuery;
import org.descartae.android.R;

import org.descartae.android.TypeOfWasteQuery;
import org.descartae.android.adapters.FacilityListAdapter;
import org.descartae.android.networking.NetworkingConstants;
import org.descartae.android.networking.apollo.errors.RegionNotSupportedError;
import org.descartae.android.presenter.facility.FacilityListPresenter;
import org.descartae.android.view.activities.FacilityActivity;
import org.descartae.android.view.events.EventHideLoading;
import org.descartae.android.view.events.EventShowLoading;
import org.descartae.android.view.utils.SimpleDividerItemDecoration;
import org.descartae.android.view.viewholder.FacilityViewHolder;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class FacilitiesFragment extends Fragment implements OnMapReadyCallback {

    private FacilityListAdapter facilityListAdapter;

    @Inject FacilityListPresenter presenter;

    @Inject EventBus eventBus;

    @BindView(R.id.container)
    public CoordinatorLayout coordinatorLayout;

    @BindView(R.id.list)
    public RecyclerView recyclerView;

    @BindView(R.id.bottom_sheet)
    public View bottomSheetList;

    @BindView(R.id.bottom_sheet_detail)
    public View bottomSheetDetail;

    @BindView(R.id.loading)
    public View mLoading;

    @BindView(R.id.filter_empty)
    public View mFilterEmpty;

    private FacilityViewHolder facilityViewHolder;

    private BottomSheetBehavior<View> behaviorDetail;
    private BottomSheetBehavior<View> behaviorList;
    private FacilitiesQuery.Item mItemSelected;

    private MapFragment mMapFragment;
    private GoogleMap mMap;

    private List<TypeOfWasteQuery.TypesOfWaste> typesOfWasteData;
    private String[] typesOfWasteTitle;
    private Integer[] selectedTypesIndices = {};

    public FacilitiesFragment() {
    }

    public static FacilitiesFragment newInstance() {
        FacilitiesFragment fragment = new FacilitiesFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {

            case R.id.action_filter:

                if ( ! presenter.haveCurrentLocation()) {
                    return false;
                }

                if (typesOfWasteTitle == null || typesOfWasteTitle.length <= 0) {
                    Log.d("Filter", "No Types");
                    return false;
                }

                new MaterialDialog.Builder(getActivity())
                        .title(R.string.title_filter)
                        .items(typesOfWasteTitle)
                        .itemsCallbackMultiChoice(selectedTypesIndices, getCallbackMultiChoiceFilter())
                        .positiveText(R.string.action_filter)
                        .show();

                return true;
        }
        return false;
    }

    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        /**
         * Init Dagger
         */
        DescartaeApp.getInstance(getActivity())
                .getAppComponent()
                .inject(this);

        /**
         * Load Type of Waste in order to be fetched before click on filter option
         */
        queryTypeOfWastes();
    }

    @Override
    public void onStart() {
        super.onStart();
        eventBus.register(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        eventBus.unregister(this);
    }

    private void queryTypeOfWastes() {

        ApolloClient apolloClient = ApolloClient.builder().serverUrl(NetworkingConstants.BASE_URL).build();
        TypeOfWasteQuery typeOfWasteQuery = TypeOfWasteQuery.builder().build();
        apolloClient.query(typeOfWasteQuery).enqueue(new ApolloCall.Callback<TypeOfWasteQuery.Data>() {

            @Override
            public void onResponse(@Nonnull final Response<TypeOfWasteQuery.Data> dataResponse) {

                if (dataResponse == null) return;
                if (dataResponse.data() == null) return;

                typesOfWasteData = dataResponse.data().typesOfWaste();
                typesOfWasteTitle = new String[typesOfWasteData.size()];

                int i = 0;
                for (TypeOfWasteQuery.TypesOfWaste type : typesOfWasteData) {
                    typesOfWasteTitle[i] = type.name();
                    i++;
                }
            }

            @Override
            public void onFailure(@Nonnull ApolloException e) {

                if (e != null && e.getMessage() != null)
                    Log.e("ApolloFacilityQuery", e.getMessage());
            }
        });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_facility_list, container, false);

        ButterKnife.bind(this, view);

        mMapFragment = MapFragment.newInstance();
        getActivity().getFragmentManager().beginTransaction().replace(R.id.map, mMapFragment).commitAllowingStateLoss();

        Context context = view.getContext();
        recyclerView.addItemDecoration(new SimpleDividerItemDecoration(getActivity()));
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        facilityListAdapter = new FacilityListAdapter(getActivity(), (FacilitiesQuery.Item center) -> {

            // On Facility Item Clicked
            mItemSelected = center;
            selectFacility(center);
        });
        recyclerView.setAdapter(facilityListAdapter);

        behaviorList = BottomSheetBehavior.from(bottomSheetList);
        behaviorList.setHideable(false);
        behaviorList.setPeekHeight(getResources().getDimensionPixelOffset(R.dimen.facilities_peek_height));

        behaviorDetail = BottomSheetBehavior.from(bottomSheetDetail);
        behaviorDetail.setBottomSheetCallback(getCallbackBottomSheet());
        behaviorDetail.setHideable(true);

        return view;
    }

    private void selectFacility(FacilitiesQuery.Item center) {

        // Fill Item Detail
        facilityViewHolder = new FacilityViewHolder(bottomSheetDetail);
        facilityViewHolder.mItem = center;
        facilityViewHolder.setCurrentLocation(presenter.getCurrentLocation());
        facilityViewHolder.fill();

        // Show BottomSheetDetail
        bottomSheetDetail.setVisibility(View.VISIBLE);
        behaviorDetail.setState(BottomSheetBehavior.STATE_EXPANDED);

        // List Collapse and GONE
        behaviorList.setState(BottomSheetBehavior.STATE_COLLAPSED);
        bottomSheetList.setVisibility(View.GONE);

        // Move Map
        if (mMap != null) {

            mMap.clear();

            LatLng latlng = new LatLng(
                    mItemSelected.location().coordinates().latitude(),
                    mItemSelected.location().coordinates().longitude()
            );

            mMap.moveCamera(CameraUpdateFactory.newLatLng(latlng));
            mMap.addMarker(
                    new MarkerOptions()
                            .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_pin))
                            .position(latlng)
                            .title(mItemSelected.name()));
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {

        mMap = googleMap;

        fillMapMarkers();

        mMap.setOnMarkerClickListener((Marker marker) -> {
            marker.setIcon(BitmapDescriptorFactory.fromResource(R.drawable.ic_pin));
            selectFacility(marker.getTitle());
            return false;
        });
    }

    private void selectFacility(String name) {
        List<FacilitiesQuery.Item> centers = facilityListAdapter.getCenters();
        if (centers != null && centers.size() > 0) {
            for (FacilitiesQuery.Item facility : centers) {
                if (facility.name().equals(name)) {
                    mItemSelected = facility;
                    selectFacility(facility);
                    break;
                }
            }
        }
    }

    private void fillMapMarkers() {

        mMap.clear();

        // Add pins to map
        List<FacilitiesQuery.Item> facilities = facilityListAdapter.getCenters();
        if (facilities != null && facilities.size() > 0) {

            for (FacilitiesQuery.Item facility : facilities) {

                LatLng latlng = new LatLng(facility.location().coordinates().latitude(), facility.location().coordinates().longitude());
                mMap.addMarker(
                        new MarkerOptions()
                                .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_places_map))
                                .position(latlng)
                                .snippet(facility.location().address())
                                .title(facility.name()));
            }
        }

        // Move camera
        Location currentLocation = presenter.getCurrentLocation();

        if (currentLocation != null && mMap != null) {
            LatLng latlng = new LatLng(currentLocation.getLatitude(), currentLocation.getLongitude());
            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latlng, 13));
        }
    }

    @NonNull
    private MaterialDialog.ListCallbackMultiChoice getCallbackMultiChoiceFilter() {
        return (MaterialDialog dialog, Integer[] which, CharSequence[] text) -> {

            selectedTypesIndices = which;

            List<String> selected = new ArrayList<>();
            for (Integer index : which) {
                selected.add(typesOfWasteData.get(index)._id());
            }

            // Clear List
            facilityListAdapter.setCenters(null);
            facilityListAdapter.notifyDataSetChanged();

            // Refresh MapMarkers
            fillMapMarkers();

            presenter.setFilterTypesID(selected);
            presenter.requestFacilities();

            return true;
        };
    }

    @NonNull
    private BottomSheetBehavior.BottomSheetCallback getCallbackBottomSheet() {
        return new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {

                if (newState == BottomSheetBehavior.STATE_HIDDEN) {

                    // List VISIBLE and Expand
                    bottomSheetList.setVisibility(View.VISIBLE);
                    behaviorList.setState(BottomSheetBehavior.STATE_EXPANDED);

                    mItemSelected = null;

                    // Reset Map Pins
                    fillMapMarkers();
                }
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                // React to dragging events
            }
        };
    }

    @OnClick(R.id.action_detail)
    public void onActionDetail() {

        if (mItemSelected == null) return;

        Intent intent = new Intent(getActivity(), FacilityActivity.class);
        intent.putExtra(FacilityActivity.ARG_ID, mItemSelected._id());
        startActivity(intent);
    }

    @OnClick(R.id.action_go)
    public void onActionGo() {

        if (mItemSelected == null) return;

        String uri = "geo:" + mItemSelected.location().coordinates().latitude() + ","
                + mItemSelected.location().coordinates().longitude() + "?q=" + mItemSelected.location().coordinates().latitude()
                + "," + mItemSelected.location().coordinates().longitude();
        startActivity(new Intent(android.content.Intent.ACTION_VIEW, Uri.parse(uri)));
    }

    @OnClick(R.id.action_clear_filter)
    public void onClearFilters() {
        selectedTypesIndices = new Integer[]{};

        mFilterEmpty.setVisibility(View.GONE);

        // Clear List
        facilityListAdapter.setCenters(null);
        facilityListAdapter.notifyDataSetChanged();

        // Refresh MapMarkers
        fillMapMarkers();

        presenter.setFilterTypesID(null);
        presenter.requestFacilities();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void eventHideLoading(EventHideLoading event) {
        mLoading.setVisibility(View.GONE);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void eventShowLoading(EventShowLoading event) {
        mLoading.setVisibility(View.VISIBLE);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void renderFacilities(FacilitiesQuery.Facilities facilities) {

        if (facilities == null && presenter.hasFilterType()) {

            /**
             * If no facilities return with filter
             */
            mFilterEmpty.setVisibility(View.VISIBLE);

        } else if (facilities != null) {

            /**
             * If have facilities
             */
            facilityListAdapter.setCenters(facilities.items());
            facilityListAdapter.setCurrentLocation(presenter.getCurrentLocation());
            facilityListAdapter.notifyDataSetChanged();

            mMapFragment.getMapAsync(FacilitiesFragment.this);
        } else {

            /**
             * If no facilities and no filter
             * @deprecated the server is already checkin this situation as Error
             */
            eventBus.post(new RegionNotSupportedError());
        }
    }

    public boolean isBottomSheetOpen() {
        return behaviorDetail.getState() == BottomSheetBehavior.STATE_EXPANDED;
    }

    public void closeBottomSheet() {
        behaviorDetail.setState(BottomSheetBehavior.STATE_HIDDEN);
    }

}

File: app/src/main/java/org/descartae/android/view/fragments/facility/FacilitiesFragment.kt
package org.descartae.android.view.fragments.facility

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapFragment
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.firebase.appindexing.Action
import com.google.firebase.appindexing.FirebaseAppIndex
import com.google.firebase.appindexing.FirebaseUserActions
import com.google.firebase.appindexing.Indexable
import com.google.firebase.appindexing.builders.Actions
import com.google.firebase.appindexing.builders.Indexables
import kotlinx.android.synthetic.main.filter_empty_view.action_clear_filter
import kotlinx.android.synthetic.main.filter_empty_view.filter_empty
import kotlinx.android.synthetic.main.fragment_facility_list.action_detail
import kotlinx.android.synthetic.main.fragment_facility_list.action_go
import kotlinx.android.synthetic.main.fragment_facility_list.bottom_sheet
import kotlinx.android.synthetic.main.fragment_facility_list.bottom_sheet_detail
import kotlinx.android.synthetic.main.fragment_facility_list.list
import kotlinx.android.synthetic.main.fragment_facility_list.loading
import org.descartae.android.DescartaeApp
import org.descartae.android.FacilitiesQuery
import org.descartae.android.R
import org.descartae.android.adapters.FacilityListAdapter
import org.descartae.android.networking.apollo.errors.RegionNotSupportedError
import org.descartae.android.presenter.facility.FacilityListPresenter
import org.descartae.android.presenter.typeofwaste.TypeOfWastePresenter
import org.descartae.android.view.activities.FacilityActivity
import org.descartae.android.view.events.EventHideLoading
import org.descartae.android.view.events.EventShowLoading
import org.descartae.android.view.utils.SimpleDividerItemDecoration
import org.descartae.android.view.viewholder.FacilityViewHolder
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.ArrayList
import javax.inject.Inject

class FacilitiesFragment : Fragment(), OnMapReadyCallback {

  private var facilityListAdapter: FacilityListAdapter? = null

  @Inject
  lateinit var presenter: FacilityListPresenter

  @Inject
  lateinit var presenterTypeWaste: TypeOfWastePresenter

  @Inject
  lateinit var eventBus: EventBus

  private var behaviorDetail: BottomSheetBehavior<View>? = null
  private var behaviorList: BottomSheetBehavior<View>? = null
  private var mItemSelected: FacilitiesQuery.Item? = null

  private var mMapFragment: MapFragment? = null
  private var mMap: GoogleMap? = null

  private var selectedTypesIndices = arrayOf<Int>()

  private val callbackBottomSheet: BottomSheetBehavior.BottomSheetCallback
    get() = object : BottomSheetBehavior.BottomSheetCallback() {
      override fun onStateChanged(bottomSheet: View, newState: Int) {

        if (newState == BottomSheetBehavior.STATE_HIDDEN) {
          bottom_sheet.visibility = View.VISIBLE
          behaviorList!!.state = BottomSheetBehavior.STATE_EXPANDED


          mItemSelected?.let {
            FirebaseUserActions.getInstance().end(getFacilityViewAction(it))
          }

          mItemSelected = null
          fillMapMarkers()
        }
      }

      override fun onSlide(bottomSheet: View, slideOffset: Float) {}
    }

  val isBottomSheetOpen: Boolean get() = behaviorDetail!!.state == BottomSheetBehavior.STATE_EXPANDED

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setHasOptionsMenu(true)
  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {

    when (item!!.itemId) {

      R.id.action_filter -> {

        if (!presenter.haveCurrentLocation()) {
          return false
        }

        if (!presenterTypeWaste.isTypesLoaded()) {
          Log.d("Filter", "No Types")
          return false
        }

        if (activity == null) return false

        MaterialDialog.Builder(activity!!)
            .title(R.string.title_filter)
            .items(*presenterTypeWaste.typesOfWasteTitle!!)
            .itemsCallbackMultiChoice(selectedTypesIndices
            ) { _, which, _ ->
              selectedTypesIndices = which

              val selected = ArrayList<String>()
              for (index in which) {
                presenterTypeWaste.getTypeId(index)?.let { selected.add(it) }
              }
              facilityListAdapter!!.centers = null
              facilityListAdapter!!.notifyDataSetChanged()
              fillMapMarkers()

              presenter.setFilterTypesID(selected)
              presenter.requestFacilities()

              true
            }
            .positiveText(R.string.action_filter)
            .show()

        return true
      }
    }
    return false
  }

  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)

    /*
     * Init Dagger
     */
    (activity!!.applicationContext as DescartaeApp).component.inject(this)

    presenter.requestLocation()
    presenterTypeWaste.requestTypeOfWastes()

    action_go.setOnClickListener {

      mItemSelected?.location()?.coordinates()?.let {
        val uri = ("geo:${it.latitude()},${it.longitude()}?q=${it.latitude()},${it.longitude()}")
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(uri)))
      }
    }

    action_clear_filter.setOnClickListener {

      selectedTypesIndices = arrayOf()

      filter_empty.visibility = View.GONE

      // Clear List
      facilityListAdapter!!.centers = null
      facilityListAdapter!!.notifyDataSetChanged()

      // Refresh MapMarkers
      fillMapMarkers()

      presenter.setFilterTypesID(null)
      presenter.requestFacilities()
    }

    action_detail.setOnClickListener {

      mItemSelected?.let {
        val intent = Intent(activity, FacilityActivity::class.java)
        intent.putExtra(FacilityActivity.ARG_ID, it._id())
        startActivity(intent)
      }
    }

    list.addItemDecoration(SimpleDividerItemDecoration(context!!))
    list.layoutManager = LinearLayoutManager(context)
    facilityListAdapter = FacilityListAdapter { center: FacilitiesQuery.Item ->

      // On Facility Item Clicked
      mItemSelected = center
      selectFacility(center)
    }
    list.adapter = facilityListAdapter

    behaviorList = BottomSheetBehavior.from(bottom_sheet)
    behaviorList!!.isHideable = false
    behaviorList!!.peekHeight = resources.getDimensionPixelOffset(R.dimen.facilities_peek_height)

    behaviorDetail = BottomSheetBehavior.from(bottom_sheet_detail)
    behaviorDetail!!.setBottomSheetCallback(callbackBottomSheet)
    behaviorDetail!!.isHideable = true
  }

  override fun onResume() {
    super.onResume()
    eventBus.register(this)
  }

  override fun onPause() {
    super.onPause()
    eventBus.unregister(this)
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?): View? {
    val view = inflater.inflate(R.layout.fragment_facility_list, container, false)
    mMapFragment = MapFragment.newInstance()
    activity!!.fragmentManager.beginTransaction().replace(R.id.map,
        mMapFragment).commitAllowingStateLoss()
    return view
  }

  private fun selectFacility(center: FacilitiesQuery.Item) {

    // Fill Item Detail
    val facilityViewHolder = FacilityViewHolder(bottom_sheet_detail)
    facilityViewHolder.mItem = center
    facilityViewHolder.setCurrentLocation(presenter.currentLocation)
    facilityViewHolder.fill()

    // Show BottomSheetDetail
    indexingFacility(center)
    FirebaseUserActions.getInstance().start(getFacilityViewAction(center))

    bottom_sheet_detail.visibility = View.VISIBLE
    behaviorDetail!!.state = BottomSheetBehavior.STATE_EXPANDED

    // List Collapse and GONE
    behaviorList!!.state = BottomSheetBehavior.STATE_COLLAPSED
    bottom_sheet.visibility = View.GONE

    // Move Map
    if (mMap != null) {

      mMap!!.clear()

      val latlng = LatLng(
          mItemSelected!!.location().coordinates().latitude(),
          mItemSelected!!.location().coordinates().longitude()
      )

      mMap!!.moveCamera(CameraUpdateFactory.newLatLng(latlng))
      mMap!!.addMarker(
          MarkerOptions()
              .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_pin))
              .position(latlng)
              .title(mItemSelected!!.name()))
    }
  }

  override fun onMapReady(googleMap: GoogleMap) {

    mMap = googleMap

    fillMapMarkers()

    mMap!!.setOnMarkerClickListener { marker: Marker ->
      marker.setIcon(BitmapDescriptorFactory.fromResource(R.drawable.ic_pin))
      selectFacility(marker.title)
      false
    }
  }

  private fun selectFacility(name: String) {
    val centers = facilityListAdapter!!.centers
    if (centers != null && centers.size > 0) {
      for (facility in centers) {
        if (facility.name() == name) {
          mItemSelected = facility
          selectFacility(facility)
          break
        }
      }
    }
  }

  private fun fillMapMarkers() {

    mMap!!.clear()

    // Add pins to map
    val facilities = facilityListAdapter!!.centers
    if (facilities != null && facilities.isNotEmpty()) {

      facilities.forEach { facility ->
        val latlng = LatLng(facility.location().coordinates().latitude(),
            facility.location().coordinates().longitude())
        mMap!!.addMarker(
            MarkerOptions()
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_places_map))
                .position(latlng)
                .snippet(facility.location().address())
                .title(facility.name()))
      }
    }

    // Move camera
    val currentLocation = presenter.currentLocation

    if (currentLocation != null && mMap != null) {
      val latLng = LatLng(currentLocation.latitude, currentLocation.longitude)
      mMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 13f))
    }
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  fun eventHideLoading(event: EventHideLoading) {
    loading.visibility = View.GONE
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  fun eventShowLoading(event: EventShowLoading) {
    loading.visibility = View.VISIBLE
  }

  @Subscribe(threadMode = ThreadMode.MAIN)
  fun renderFacilities(facilities: FacilitiesQuery.Facilities?) {

    val hasFacilities = facilities?.items() != null && facilities.items()!!.size > 0

    if (!hasFacilities && presenter.hasFilterType()) {

      /*
       * If no facilities return with filter
       */
      filter_empty.visibility = View.VISIBLE

    } else if (hasFacilities) {

      /*
       * If have facilities
       */
      facilityListAdapter!!.centers = facilities!!.items()
      facilityListAdapter!!.currentLocation = presenter.currentLocation
      facilityListAdapter!!.notifyDataSetChanged()

      mMapFragment!!.getMapAsync(this@FacilitiesFragment)
    } else {

      /*
        If no facilities and no filter
        @deprecated the server is already checkin this situation as Error
       */
      eventBus.post(RegionNotSupportedError())
    }
  }

  fun closeBottomSheet() {
    behaviorDetail!!.state = BottomSheetBehavior.STATE_HIDDEN
  }

  private fun indexingFacility(facility: FacilitiesQuery.Item) {
    val facilityToUpdate = Indexables.placeBuilder()
        .setName(facility.name())
        .setUrl(getString(R.string.deeplink_uri, facility._id()))
        .setDescription(facility.location().address())
        .build()

    val task = FirebaseAppIndex.getInstance().update(facilityToUpdate)
    task.addOnSuccessListener {
      Log.d(
          "App Indexing",
          "App Indexing API: Successfully added ${facility.name()} to index"
      )
    }
    task.addOnFailureListener {
      Log.e(
          "App Indexing",
          "App Indexing API: Failed to add ${facility.name()} to index. ${it.message}"
      )
    }
  }

  private fun getFacilityViewAction(facility: FacilitiesQuery.Item): Action? {
    return Actions.newView(facility.name(), getString(R.string.deeplink_uri, facility._id()))
  }

  companion object {

    fun newInstance(): FacilitiesFragment {
      val fragment = FacilitiesFragment()
      val args = Bundle()
      fragment.arguments = args
      return fragment
    }
  }
}